docs/excel/custom-functions-runtime.md

```diff
@@ -0,0 +1,149 @@
+    }
+}
+
+```
+
+## WebSockets
+
+WebSockets is a networking protocol which creates real-time communication between a server and one or more clients. It is often used for chat applications because it allows you to read and write text simultaneously.  
+
+As you can see in the sample below, custom functions can use WebSockets. In this example, the WebSocket logs any message sent to it.  
+
+```js
+const ws = new WebSocket('wss://bundles.office.com');
+ws.onmessage = (message) => {
+    console.log(`Received: ${message}`);
+};
+ws.onerror = (error) => {
+    console.err(`Failed: ${error}`);
+}
+```
+
+## AsyncStorage
+
+AsyncStorage is a key-value storage system that can be used to store authentication tokens or used for:
+
+- Persistent
+- Unencrypted
+- Asynchronous
+
+Additionally, AsyncStorage is available globally to all parts of your add-in. For custom functions, Async Storage is exposed as a global object. For other parts of your add-in, such as taskpanes and other elements which utilize the typical WebView runtime, AsyncStorage is exposed through `Office.Runtime`.
+
+ Methods available on AsyncStorage include getItem, setItem, removeItem, clear, getAllKeys, flushGetRequests, multiGet, multiSet, and multiRemove. At this time, mergeItem and multiMerge are not supported methods.
+
+Each add-in has its own storage partition, with a default of 5MB of storage.  
+
+The code sample below illustrates the process of getting an item from AsyncStorage:
+
+```js
+_goGetData = async () => {
+    try {
+        const value = await AsyncStorage.getItem('toDoItem');
+        if (value !== null) {
+            //data exists and you can do something with it here
+            }
+        } catch (error) {
+            //handle errors here
+        }
+    }
+}
+```
+
+## Dialog API
+
+The Dialog API allows you to require user authentication through an outside resource, such as Google or Facebook, before they can use your function. The Dialog API allows you to open a dialog box which prompts user sign-in.  
+
+The code sample below illustrates the use of the Dialog API’s `displayWebDialog()` method.  
+
+```ts
+// Get auth token before calling my service, a hypothetical API which will deliver a stock price based on stock ticker string, such as "MSFT"
+async function getStock(ticker) {
+    const token = await getToken();
+    let data = await (await fetch(https://myservice.com/?token=token&ticker= + ticker).json());
+    return data.price;
+}
+
+async function getToken() {
+    if (_cachedToken) {
+        return _cachedToken;
+    } else {
+        return await getTokenViaDialog_AsPromise();
+    }
+}
+  
+// Function to display dialog window
+function getTokenViaDialog_AsPromise() {
+    return new Promise ((resolve, reject) => {
+        displayWebDialog("https://www.auth.com/", {
+           height: ’50’,
+           width: ’50%’,
+           hideTitle: true,
+           onMessage: (message, dialog) => {
+                let json = JSON.parse(message);
```

Comment:
Another place where a "let" should be a "const". Please check all the code examples for this.